newIDE/app/src/ObjectsList/index.js

```diff
@@ -637,94 +858,390 @@ const ObjectsList = React.forwardRef<Props, ObjectsListInterface>(
+        forceUpdateList,
+        objectsContainer,
+        selectObjectFolderOrObjectWithContext,
+        editName,
+      ]
     );
-    const eventsFunctionsExtensionWriter = eventsFunctionsExtensionsState.getEventsFunctionsExtensionWriter();
 
     const renderObjectMenuTemplate = React.useCallback(
-      (i18n: I18nType) => (
-        objectWithContext: ObjectWithContext,
-        index: number
-      ) => {
-        const { object } = objectWithContext;
+      (i18n: I18nType) => (item: TreeViewItem, index: number) => {
+        if (item.isRoot || item.isPlaceholder) return [];
+        const { objectFolderOrObject, global } = item;
+
+        const container = global ? project : objectsContainer;
+        let folderAndPathsInContainer = enumerateFoldersInContainer(container);
+        folderAndPathsInContainer.unshift({
+          path: i18n._(t`Root folder`),
+          folder: container.getRootFolder(),
+        });
+        if (objectFolderOrObject.isFolder()) {
+          folderAndPathsInContainer = folderAndPathsInContainer.filter(
+            folderAndPath =>
+              !folderAndPath.folder.isADescendantOf(objectFolderOrObject) &&
+              folderAndPath.folder !== objectFolderOrObject
+          );
+          return [
+            {
+              label: getPasteLabel(i18n, {
+                isGlobalObject: item.global,
+                isFolder: true,
+              }),
+              enabled: Clipboard.has(CLIPBOARD_KIND),
+              click: () => paste(item, true),
+            },
+            {
+              label: i18n._(t`Rename`),
+              click: () => editName(item),
+              accelerator: getShortcutDisplayName(
+                preferences.values.userShortcutMap['RENAME_SCENE_OBJECT'] ||
+                  defaultShortcuts.RENAME_SCENE_OBJECT
+              ),
+            },
+            {
+              label: i18n._(t`Delete`),
+              click: () => deleteObjectFolderOrObjectWithContext(item),
+              enabled: objectFolderOrObject.getChildrenCount() === 0,
+              accelerator: 'Backspace',
+            },
+            {
+              label: i18n._('Move to folder'),
+              submenu: folderAndPathsInContainer.map(({ folder, path }) => ({
+                label: path,
+                enabled: folder !== objectFolderOrObject.getParent(),
+                click: () => {
+                  objectFolderOrObject
+                    .getParent()
+                    .moveObjectFolderOrObjectToAnotherFolder(
+                      objectFolderOrObject,
+                      folder,
+                      0
+                    );
+                  onObjectModified(true);
+                  if (treeViewRef.current)
+                    treeViewRef.current.openItems([
+                      getTreeViewItemId({
+                        objectFolderOrObject: folder,
+                        global: item.global,
+                      }),
+                    ]);
```

Comment:
Could almost be extracted as a function, it's a bit of logic buried inside a menu.